core/src/runtime/opaque.rs

```diff
@@ -25,6 +32,13 @@ pub(crate) struct Opaque<'js> {
     /// The user provided interrupt handler, if any.
     interrupt_handler: UnsafeCell<Option<InterruptHandler>>,
 
+    /// The class id for rust classes.
+    class_id: qjs::JSClassID,
+    /// The class id for rust classes which can be called.
+    callable_class_id: qjs::JSClassID,
+
+    prototypes: UnsafeCell<HashMap<*const VTable, Option<Object<'js>>>>,
```

Comment:
As we'll always add prototypes and never remove them we can use TinyVec instead and store the index in the VTable. Something  like this (psudocode):
```rust
prototypes: Unsafecell<TinyVec<[(*const VTable, Option<Object<'js>>); 1024]>>
```
```rust
pub fn get_or_insert_prototype<C: JsClass<'js>>(
        &self,
        ctx: &Ctx<'js>,
    ) -> Result<Option<Object<'js>>, Error> {
        unsafe {
            let vtable = VTable::get::<C>();
            let prototypes =  (*self.prototypes.get());
            if let Some(index) = vtable.index {
                return prototypes[index].1
            }
            let next_index = prototypes.length();
            vtable.index = Some(next_index);
            prototypes.push((vtable,C::prototype(ctx)?)
        }
    }
```

Reply:
Just occurred to me that this might also be a bit more complex and not feasible as we use the same vtable for many runtimes.

Reply:
So think i figured out a great way to solve this without hashtable:

1. Each vtable gets an additional auto incrementing index_id (derived form an atomic u32)
2. `prototypes` are stored as a `Vec` (or `TinyVec`)
3. Proto objects are stored and acceced via the array index of index_id for the vtable

Fast, constant time lookup, no hashing or collitions